Engineering Principles (#6)
MetaMask/contributor-docs: policies/engineering-principles.md

### The Principle of Code Purity

Whenever possible prefer pure, functional code over imperative code. Pure code is free of side effects, and will return the same output given the same inputs. Imperative code is code that produces side effects, such as mutation or state changes. Imperative code is not avoidable but we should strive to keep as much of our application logic pure as possible. 

### The Principle of Incremental Change

To minimize the risk to the stability and security of the application, changes to the code should be incremental and as small as possible. Large changes to the codebase become difficult to review effectively and make the commit history less itemized. Furthermore, incremental change promotes better test coverage and makes it easier to require tests to cover the introduced changes.

### The Principle of Incremental Decentralization

> We may sometimes ship with centralized services by default, but if we must, users should be able to opt out of any of these during onboarding, ideally easily swapping them out for decentralized alternatives. When we build features that rely upon centralized services we must build them in such a way where the experience isn't completely broken when the service is disrupted. We are trying to build a decentralized future and we must make choices that move us directionally towards those goals.

Author: I agree its a reliability issue, however it bears mentioning in the architectural principle of decentralization. If we are going to make a choice to add another centralized service to the code, we must design that code with the ability to opt out of using it and therefore the decisions we make (in order to also provide a reliable application) must be cognizant of this fact.

### The Principle of Self-Soverign Identity

We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right we must treat those features as opt in, not opt out. When we do build features that make user data known to us we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity protecting way. This principle must be extended to cover all third party code as well.

### The Principle of Auditability

> All code contributions must be reviewed thoroughly for code quality, usability and adherence to these principles. Code contributions must be kept as small as possible (see Principle of Incremental Change) to promote auditability. Avoid long running feature branches because they are not auditable, and pose a risk to the stability and security of the product. We should engage in self-audit of code that we contribute, as well as actively engage in code review of others' pull requests to ensure best practices, and engineering principles are reflected in the work.

Reviewer: This might be clearer if we frame it as a commitment to ensuring our work is auditable, rather than as a commitment to actively engage in review.

e.g. We could start by stating a commitment to transparency (should this be a value as well? :thinking: ), and/or making a connection between security and the codebase being publicly accessible.

If our codebase:
* is easy to read
* follows consistent conventions
* maintains a high standard for code usability/quality
* has a public record of why changes are made (the git/PR history)

then our codebase will be more maintainable as a direct result, and it will be easier for us to uphold our commitments to security and stability. And users can trust us less by directly auditing the code themselves, or hiring third parties to audit our codebase (especially important as part of a postmortem when things go wrong).

We've paid for third-party audits before, so this isn't even a theoretical concern. This would directly benefit us even in that sense.

This framing of auditability makes more sense as a "principle" to me. Whereas practices like code review, standards for accepting contributions, practices like long-running feature branches, those are.... well, they're details, means to an end. Of course we want tall of those things but we want to maintain a high degree of team autonomy to adjusting their development practices to best suit their own needs, and I don't think we need to commit to them at this level.

Reviewer: Let me know if that makes sense, and I can try to produce a coherent draft. I tried just now but had to give up and revert to the bulleted list.

Reviewer 2: I agree that this reads more like a list of rules than a principle that those rules could be based on. I think reframing this in terms of transparency a lot more sense.

Reviewer 2: That said, we could also change this in a future PR.

Author: I took a stab at words to capture your intent here @Gudahtt , but i'm not 100% on it. Could you take a look and make suggestions? https://github.com/MetaMask/contributor-docs/pull/6/commits/0dcb4cc296adb46eca7570ef302ccc82e4b97688

### The Principle of "Be Methodical and Fix Things"

> We should prioritize fixing bugs and improving stability over creating new features whenever possible. All code generates new bugs and builds technical debt, this is inevitable. However, if we focus on building our features with stability in mind and throw out the "Move Fast and Break Things" mentality of most web 2.0 applications we can minimize the amount of bugs we introduce. Furthermore, whenever bugs are known to exist we should expend the necessary energy to remedy them. 

Reviewer: Nit: This is fine as-is, but, I'm sure we could do better here. This doesn't present a convincing argument for _why_ we should "prioritize fixing bugs and improving stability", or paint a clear picture of what that entails.

On the "why", just saying that we can minimize bugs is underselling it. By being methodical and fixing things, we can maintain a more consistent developer velocity by minimizing disruptions caused by bugs, and we can improve velocity in the long-term by taking steps to avoid the buildup of technical debt. We can frame this principle as a commitment to _sustainable development_; to setting a pace that we can maintain years out, not just focusing on the current quarterly goals.

Then there's the user impact angle. In a sense, being methodical is being considerate of our users and the impact our work might have on them. And fixing things is a commitment to product quality.

On the "how", prioritizing fixing bugs and improving stability sounds great to me. Maybe we can say something more about the planning stage? On anticipating problems before they occur, having well thought out designs. Or maybe being methodical could refer to our efforts to maintain an auditable codebase. I'm less sure on this part. I guess we don't really need to spell out the "how" for any of these principles though, not here anyway.

Reviewer 2: I agree that this doesn't go far enough in explaining the *why* (this is sort of a problem with all of these principles, but anyway). However, I agree with the conclusion of this paragraph. My vote is to expand this in a future PR.

Author: @mcmire We should be aiming to cover the why in these, at least at the surface level. I think fully covering the how is best left for supporting documents. That said, if there are principles that do not go far enough to cover the why please don't hesitate to point them out so we can get to a place that covers the why enough to get us into the supporting documents.

Reviewer 2: Sounds good!

Author: I will take another stab at rewording this in a subsequent PR, we are at least agreed that it should be a principle we just need to do a better job detailing it.

> ### The Principle of Relability

Reviewer: Typo:

```suggestion
### The Principle of Reliability
```

> We provide a pleathora of features in our applications that users expect to be supported, and that they rely upon for their day to day activities. We must handle feature deprecation gracefully, and inform users well in advance that a feature is being sunsetted or replaced. If we remove features or experiences from our product without warning it will alienate our users and cause them to seek out other solutions. 

Reviewer 2: Typo: 

``` suggestion
We provide a plethora of features in our applications that users expect to be supported, and that they rely upon for their day to day activities. We must handle feature deprecation gracefully, and inform users well in advance that a feature is being sunsetted or replaced. If we remove features or experiences from our product without warning it will alienate our users and cause them to seek out other solutions. 
```

> We also provide external facing APIs that app developers utilize to build their products. Every change we make to thes APIs must be analyzed for potentially breaking functionality of the apps in our ecosystem. When we make breaking changes we are creating work for countless other organizations and individuals, and only by their good graces will they make the requested changes. We must view these breaking changes as a potential reset of our momentum within the ecosystem. Breaking changes should only be made for critical reasons of security or stability that cannot be made by any other means.

Reviewer 2: Typo:

``` suggestion
We also provide external facing APIs that app developers utilize to build their products. Every change we make to these APIs must be analyzed for potentially breaking functionality of the apps in our ecosystem. When we make breaking changes we are creating work for countless other organizations and individuals, and only by their good graces will they make the requested changes. We must view these breaking changes as a potential reset of our momentum within the ecosystem. Breaking changes should only be made for critical reasons of security or stability that cannot be made by any other means.
```

### The Principle of Incremental Change

To minimize the risk to the stability and security of the application, changes to the code should be incremental and as small as possible. Large changes to the codebase become difficult to review effectively and make the commit history less itemized. Furthermore, incremental change promotes better test coverage and makes it easier to require tests to cover the introduced changes.

### The Principle of Incremental Decentralization

We are trying to build a decentralized future and we must make choices that move us directionally towards those goals. This means that users should be able to swap centralized services out for decentralized alternatives. Additionally, when we build features that rely upon centralized services we must build them in such a way where the experience isn't completely broken when the service is disrupted. 

### The Principle of Self-Sovereign Identity

> We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right we must treat those features as opt in, not opt out. When we do build features that make user data known to us we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity protecting way. This principle must be extended to cover all third party code as well.

Reviewer 3: ```suggestion
We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right, we must treat those features as opt-in, not opt-out. When we do build features that make user data known to us, we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity-protecting way. This principle must be extended to cover all third-party code as well.
```

### The Principle of Auditability

> As we are committed to providing a high level of security and privacy within our product, we must also maintain a auditable, publically accessible, and transparent codebase. We musn't ask our users to trust our products implicitly, but instead write code that explicitly illustrates our values. In order for our codebase to be easily auditable, our code must have the following qualities: 

Reviewer 3: ```suggestion
As we are committed to providing a high level of security and privacy within our product, we must also maintain an auditable, publically accessible, and transparent codebase. We mustn't ask our users to trust our products implicitly, but instead, write code that explicitly illustrates our values. In order for our codebase to be easily auditable, our code must have the following qualities: 
```

1. Readability - Our code should be easily read, and be free of obfuscation or subterfuge.
2. Consistency - Our code should follow consistent standards to assist auditors in tracing code paths across multiple repositories or code from varying authors. 
3. Usability - Our code must maintain a standard of high quality and usability, and reflect the qualities listed in the values section above.
4. Traceability - Our code should have a public record of changes made, with details about why the change was made.

### The Principle of "Be Methodical and Fix Things"

We should prioritize fixing bugs and improving stability over creating new features whenever possible. All code generates new bugs and builds technical debt, this is inevitable. However, if we focus on building our features with stability in mind and throw out the "Move Fast and Break Things" mentality of most web 2.0 applications we can minimize the amount of bugs we introduce. Furthermore, whenever bugs are known to exist we should expend the necessary energy to remedy them. 

### The Principle of Reliability

> We provide a plethora of features in our applications that users expect to be supported, and that they rely upon for their day to day activities. We must handle feature deprecation gracefully, and inform users well in advance that a feature is being sunsetted or replaced. If we remove features or experiences from our product without warning it will alienate our users and cause them to seek out other solutions. 

Reviewer 3: ```suggestion
We provide a plethora of features in our applications that users expect to be supported, and that they rely upon for their day-to-day activities. We must handle feature deprecation gracefully, and inform users well in advance that a feature is being sunsetted or replaced. If we remove features or experiences from our product without warning it will alienate our users and cause them to seek out other solutions. 
```
